### ISSUE TITLE
OuterTypeNumber violation is not suppressed with @SuppressWarnings annotation.

### ISSUE THREAD

**Issue by Fernal73** [checkstyle/checkstyle#7082]
EDITED:

$ javac Puzzle.java package-info.java
```java
@SuppressWarnings({"checkstyle:outertypenumber","checkstyle:outertypenumbercheck"})

package javapuzzles;


public class Puzzle extends Superb {
  final byte x = 1;
  final Byte y = 2;

  @Override
  public String toString() {
    return x + " : " + y;
  }

  public static void main(String... args) {
    new Puzzle();
  }
}

class Superb {
  // clang-format off
  {
    System.out.println(this);
  }
  // clang-format on
}
````
package-info.java
```java
@SuppressWarnings({"checkstyle:outertypenumber"})
package javapuzzles;
```

Compiler output:

Puzzles
warning: [options] bootstrap class path not set in conjunction with -source 10
Puzzles/javapuzzles/Puzzle.java:1: error: package annotations should be in file package-info.java
@SuppressWarnings({"checkstyle:outertypenumber"})
^

Puzzles/javapuzzles/package-info.java:1: error: package javapuzzles has already been annotated
@SuppressWarnings({"checkstyle:outertypenumber"})
^
2 errors
1 warnings

 cat config.xml 
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
<module SuppressWarningsFilter/>
  <module name="TreeWalker">
    <module name="OuterTypeNumber" />

    <module name="SuppressionXpathFilter">
      <property name="optional"
                value="true" />
      <property name="file"
                value="suppressions-xpath.xml" />
    </module>
<module name="SuppressWarningsHolder" />
    </module>
</module>
```

 $ java checkstyle-8.24-all.jar  -o output.txt -c config.xml Puzzle.java package-info.java

---------------

Describe what you expect in detail.

The checkstyle warning is expected to be suppressed in the checkstyle output. It is if placed at the package level in the class source. However, it will not compile if the package has a package-info.java file.

You'd expect the suppression to be enforced across the package for all files if placed in the package-info.java file. That does not happen.

The suppression generated using the -g option is:

```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE 
suppressions PUBLIC "-//Checkstyle//DTD SuppressionXpathFilter Experimental Configuration 1.2//EN"
"https://checkstyle.org/dtds/suppressions_1_2_xpath_experimental.dtd">
<suppressions>
  <suppress-xpath
       files="Puzzle.java"
       checks="OuterTypeNumberCheck"
       query="/PACKAGE_DEF[./IDENT[@text='javapuzzles']]"/>
</suppressions>
```
The name in the xpath file differs from the Rule name. That is the case for all rules listed in the suppressions xpath xml.

**Comment by romani** [comment]
Config is missing

**Comment by rnveach** [comment]
@Fernal73 Please update your post to include your config and the results you are experiencing.

**Comment by Fernal73** [comment]
Sure. I had to take a break. So where do I save the draft?

**Comment by rnveach** [comment]
You just need to edit your existing post above. If your using a browser, there should be 3 dots on the right hand side you can click on.
Just add your config underneath `$ cat config.xml` and your results underneath `$ java checkstyle-8.24-all.jar -c config.xml Puzzle.java`.
We need to this to be able to reproduce your issue to make sure it is valid.

**Comment by romani** [comment]
> It would help if the template included the xml wrapper so that I can just paste the rule in.

Some Checks are under Checker some under TreeWalker, you will have to update it no matter what, it is hard to make good template for all

**Comment by Fernal73** [comment]
That's fine and expected. There's a request for the draft feature for Github. 
https://github.com/isaacs/github/issues/1648

Is it the right place ?

**Comment by rnveach** [comment]
@Fernal73 
> OuterTypeNumber violation is not suppressed with @SuppressWarnings annotation
> config.xml

Your example configuration has no `SuppressWarningsFilter` to make it suppress violations by that annotation.
https://checkstyle.sourceforge.io/config_filters.html#SuppressWarningsFilter

>  -g 

This is not to be used with SuppressWarnings annotation. This is for a separate file that is given to SuppressionXpathFilter.
https://checkstyle.sourceforge.io/config_filters.html#SuppressionXpathFilter

> `@SuppressWarnings({"checkstyle:outertypenumber","checkstyle:outertypenumbercheck"})`

Please check https://checkstyle.sourceforge.io/config_filters.html#SuppressWarningsFilter for the format of the suppression.
https://checkstyle.sourceforge.io/config_filters.html#SuppressWarningsFilter_Examples
It should be just the check name and nothing else.

**Comment by Fernal73** [comment]
The config file is updated. 
Save the suppression xpath.xml generated as suppressions-xpath.xml and run checkstyle again to see if suppressions are in effect.

My full configuration is :

https://github.com/Fernal73/LearnJava/blob/master/csrules.xml

**Comment by rnveach** [comment]
Since your main issue is with SuppressWarningsFilter, your current config is incomplete.
As the documentation and examples show, you must add `SuppressWarningsHolder` for the filter work.
That and the format of the SuppressWarnings annotation is incorrect.

**Comment by Fernal73** [comment]
I've added the Holder module. 
This is the file on Github.

https://github.com/Fernal73/LearnJava/blob/master/Puzzles/javapuzzles/Puzzle.java
And this is the suppressions file:

https://github.com/Fernal73/LearnJava/blob/master/Puzzles/suppressions-xpath.xml

The java file does not have the annotation since it's not working as listed above. I think I have another class instance in that very folder that has the same error. 

What's wrong with the annotation? I'd love not to have to write up a bug. It's a bloody pain to have to do it from my phone browser .

**Comment by Fernal73** [comment]
What exactly are you implying? That your test suite is run without running the suppression annotation or suppression generation against each violation as part of the same test? That would catch these kind of errors.

**Comment by Fernal73** [comment]
What's the rationale behind having two similar checks, OneTopLevelClass and OuterTypeNumber?

**Comment by Fernal73** [comment]
The HTML documentation isn't rendering well within the mobile browser. Wasn't an issue raised for it last week?

**Comment by rnveach** [comment]
@Fernal73 Your config is invalid and you haven't tried what i suggested.

`@SuppressWarnings` has to be like `@SuppressWarnings("OuterTypeNumber")` and it has to be placed on the package as that is where the violation is being generated.

````
$ cat TestClass.java
@SuppressWarnings("OuterTypeNumber")
package javapuzzles;

public class Puzzle extends Superb {
  final byte x = 1;
  final Byte y = 2;

  @Override
  public String toString() {
    return x + " : " + y;
  }

  public static void main(String... args) {
    new Puzzle();
  }
}

class Superb {
  // clang-format off
  {
    System.out.println(this);
  }
  // clang-format on
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="SuppressWarningsFilter" />
  <module name="TreeWalker">
    <module name="OuterTypeNumber" />
    <module name="SuppressWarningsHolder" />
  </module>
</module>

$ java -jar checkstyle-8.24-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

That shows how to suppress the violation with the `@SuppressWarnings`.

**Comment by Fernal73** [comment]
On the package? That's allowed? That's the first rule I've come across that has to be placed there. The documentation doesn't state anything about it.

**Comment by Fernal73** [comment]
Having this Suppression at the top of my file breaks my Regexp header rule. 

  <module name="RegexpHeader">
    <property name="headerFile"
              value="${cs.regexp.header.file}" />
    <property name="fileExtensions"
              value="java" />
  </module>
    <module name="LineLength">
      <property name="max"
                value="100" />
      <property name="ignorePattern"
                value="^package.*|^import.*|a href|href|http://|https://|ftp://" />
    </module>

So add a SuppressWarnings annotation to the mix, eh?

**Comment by rnveach** [comment]
The documentation doesn't lay out the details, but the annotation must be placed as a parent of the violation. In this case the violation is on the package, so it must be placed around there.
I did not look into if the annotation is allowed there by the compiler. If it is not allowed, you will have to use the file suppression.

**Comment by Fernal73** [comment]
Does it provide the line number? I don't recall whether it does. The other
rule, OneTopLevelClass provides the line number. So I have a general idea
where to place it. If I recall correctly, there was a discussion related to
the same in the issues section on Github. Can these be culled to update the
documentation?

**Comment by rnveach** [comment]
Yes it provides the line number.
Please create a new issue on updating the documentation as there is too much discussion in here.
Be sure the first post contains everything on what you want updated, how and why.

**Comment by Fernal73** [comment]
Nope. If the line numbers are present, it's fine. These are technical users, not end-users. There are some checks for which suppressions are not generated. These have to be suppressed via Annotations or comments. These are the ones under the Checker module, as against TreeChecker. These could be asterisked as not supported for XML xpath suppressions.

**Comment by Fernal73** [comment]
Wait a minute. I receive a compile error when using the above annotation on the package level.

The annotation has to be placed in a package-info.java file since it exists. 

This , in turn, generates a violation of the OuterTypeNumber rule. This needs to be documented or change the annotation level to class or merge with OneTopLevelClass.

**Comment by Fernal73** [comment]
The title of the issue needs an update.

**Comment by Fernal73** [comment]
```java
@SuppressWarnings({"checkstyle:outertypenumber"})
package javapuzzles;
```

package-info.java file.

**Comment by Fernal73** [comment]
https://github.com/junit-team/junit4/issues/670

Is this useful ?

**Comment by Fernal73** [comment]
Additionally, I'm not happy with the name OuterTypeNumber. It appears to refer to number of nested classes permitted, instead of top level classes in a java file.

**Comment by romani** [comment]
@Fernal73 , too much comments .... please create issue from scratch with all required configs, outputs and definition of problem and conflicts with other tools.

**Comment by Fernal73** [comment]
I'll look into rewriting this up. It started as one error but transformed
into another.
